services/api-gateway/src/simcore_service_api_gateway/auth.py

```diff
@@ -0,0 +1,94 @@
+import logging
+from typing import Optional
+
+from fastapi import Depends, HTTPException, Security, status
+from fastapi.security import OAuth2PasswordBearer, SecurityScopes
+
+from . import crud_users as crud
+from .__version__ import api_version_prefix
+from .auth_security import get_access_token_data
+from .schemas import TokenData, User, UserInDB
+
+log = logging.getLogger(__name__)
+
+
+# Resource SERVER ----------------------------------------------
+#
+#  +--------+                               +---------------+
+#  |        |--(A)- Authorization Request ->|   Resource    |
+#  |        |                               |     Owner     | Authorization request
+#  |        |<-(B)-- Authorization Grant ---|               |
+#  |        |                               +---------------+
+#  |        |
+#  |        |                               +---------------+
+#  |        |--(C)-- Authorization Grant -->| Authorization |
+#  | Client |                               |     Server    | Token request
+#  |        |<-(D)----- Access Token -------|               |
+#  |        |                               +---------------+
+#  |        |
+#  |        |                               +---------------+
+#  |        |--(E)----- Access Token ------>|    Resource   |
+#  |        |                               |     Server    |
+#  |        |<-(F)--- Protected Resource ---|               |
+#  +--------+                               +---------------+
+#
+#                  Figure 1: Abstract Protocol Flow
+
+
+# callable with request as argument -> extracts token from Authentication header
+oauth2_scheme = OAuth2PasswordBearer(
+    tokenUrl=f"{api_version_prefix}/token",
+    scopes={
+        "me": "Read information about the current user.",
+        "projects": "Read projects.",
+        "you": "Some other scope",
+    },
+)
+
+
+async def get_current_user(
+    security_scopes: SecurityScopes, access_token: str = Depends(oauth2_scheme)
+) -> User:
+    # TODO: SecurityScopes dependnecy?? ????
+    #
+    # security_scopes is FILLED with dependant scopes. Therefore it will
+    # be filled differently depending who is calling it
+    if security_scopes.scopes:
+        authenticate_value = f'Bearer scope="{security_scopes.scope_str}"'
+    else:
+        authenticate_value = f"Bearer"
+
+    credentials_exception = HTTPException(
+        status_code=status.HTTP_401_UNAUTHORIZED,
+        detail="Could not validate credentials",
+        headers={"WWW-Authenticate": authenticate_value},
```

Comment:
the docker registry also uses this www-authenticate thingy. they changed the capitalization of the name to all small letters. not sure whether it's a problem though.